File: api/app/schemas/stats/stats_schemas.py
from pydantic import BaseModel
from typing import List
from enum import IntEnum

from app.schemas.anime import (
    StatusViewEnum,
    TipoAnimeEnum,
)
from app.schemas.manga import (
    TipoMangaEnum,
)


class TypeStatisticEnum(IntEnum):
    a_m_favs = 1  # Sera para conteo general en el caso del dashboard
    tipo_a_m = 2
    generos = 3
    studios = 4
    editorials = 5


# Conteo de anime y manga por estado de visualizacion
class StatusViewCountSchema(BaseModel):
    statusView: StatusViewEnum
    conteo: int


# Schema Conteo de tipos de anime
class StatsTipoAnimeSchema(BaseModel):
    code: TipoAnimeEnum
    conteo: int = 0
    nombre: str


# Schema conteo de tipos de manga
class StatsTipoMangaSchema(BaseModel):
    code: TipoMangaEnum
    conteo: int = 0
    nombre: str


# Schema conteo de animes y mangas por genero
class StatsGeneroSchema(BaseModel):
    id_MAL: int
    nombre: str
    conteomangas: int = 0
    conteoanimes: int = 0


# Schema de top studios de anime con mas producciones
class TopStudiosSchema(BaseModel):
    id_MAL: int
    nombre: str
    conteoanimes: int = 0


# Schema de top editoriales de manga con mas publicaciones
class TopEditorialsSchema(BaseModel):
    id_MAL: int
    nombre: str
    conteomangas: int = 0


class StatsSchema(BaseModel):
    tiposAnime: List[StatsTipoAnimeSchema] = []
    tiposManga: List[StatsTipoMangaSchema] = []
    conteoGeneros: List[StatsGeneroSchema] = []
    topStudios: List[TopStudiosSchema] = []
    topEditorials: List[TopEditorialsSchema] = []


# Schema del conteo general de animes y mangas favoritos
class FavsCountSchema(StatsSchema):
    totalAnimes: int = 0
    conteos_statusA: List[StatusViewCountSchema] = []  # Conteo por statusview de animes
    totalMangas: int = 0
    conteos_statusM: List[StatusViewCountSchema] = []  # Conteo por statusview de mangas


class ConteoGeneralSchema(StatsSchema):
    totalUsuarios: int = 0
    totalAnimes: int = 0
    totalMangas: int = 0
    totalAutoresMangas: int = 0
    totalStdAnime: int = 0
    totalEdtManga: int = 0
    totalGeneros: int = 0

File: client/src/schemas/statsSchemas.ts
import * as z from "zod";

export const StatsTipoAMZ = z.object({
     code: z.number().default(0),
     conteo: z.number().default(0),
     nombre: z.string().default("")

})

export const StatsGeneroZ = z.object({
     id_MAL: z.number().default(0),
     nombre: z.string().default(""),
     conteomangas: z.number().default(0),
     conteoanimes: z.number().default(0)
})

export const TopStudiosZ = z.object({
     id_MAL: z.number().default(0),
     nombre: z.string().default(""),
     conteoanimes: z.number().default(0)
})

export const TopEditorialsZ = z.object({
     id_MAL: z.number().default(0),
     nombre: z.string().default(""),
     conteomangas: z.number().default(0)
})

export const StatsZ = z.object({
     tiposAnime: z.array(StatsTipoAMZ).default([]),
     tiposManga: z.array(StatsTipoAMZ).default([]),
     conteoGeneros: z.array(StatsGeneroZ).default([]),
     topStudios: z.array(TopStudiosZ).default([]),
     topEditorials: z.array(TopEditorialsZ).default([]),

})

export const FullStatsZ = StatsZ.extend({
     totalUsuarios: z.number().default(0),
     totalAnimes: z.number().default(0),
     totalMangas: z.number().default(0),
     totalAutoresMangas: z.number().default(0),
     totalStdAnime: z.number().default(0),
     totalEdtManga: z.number().default(0),
     totalGeneros: z.number().default(0),
})

export const StatusViewCountZ = z.object({
     statusView: z.number().default(0),
     conteo: z.number().default(0),
})

export const FavsCountZ = StatsZ.extend({
     totalAnimes: z.number().default(0),
     conteos_statusA: z.array(StatusViewCountZ).default([]),
     totalMangas: z.number().default(0),
     conteos_statusM: z.array(StatusViewCountZ).default([]),
})

export type FavsCountSchema = z.infer<typeof FavsCountZ>
export type StatusViewCountSchema = z.infer<typeof StatusViewCountZ>
export type FullStatsSchema = z.infer<typeof FullStatsZ>